x-pack/plugins/infra/server/lib/alerting/metric_threshold/metric_threshold_executor.ts

```diff
@@ -147,8 +152,11 @@ export const createMetricThresholdExecutor = (libs: InfraBackendLibs) =>
         //   .map((result) => buildRecoveredAlertReason(formatAlertResult(result[group])))
         //   .join('\n');
       }
-      if (alertOnNoData) {
-        if (nextState === AlertStates.NO_DATA) {
+      if (alertOnNoData || alertOnGroupDisappear) {
+        if (
+          nextState === AlertStates.NO_DATA &&
+          (alertOnNoData || groups.length > 1) // Handle the possibility that the user only wants to be alerted on disappearing groups but not general No Data
+        ) {
```

Comment:
This nested ifs with || conditionals feels hard to understand. Is there a better way we could unwind these conditions at all? 

For example, it _feels_ like line 158 should be `(alertOnNoData || groups.length === 0)`, but I can't tell if I'm just parsing it wrong.

Maybe something like
```ts
// We should trigger a "no data" alert if the "Alert me if no data" option is selected OR if the "Alert me if a group stops reporting data" option is selected and no groups are returned

const shouldTriggerNoDataAlert = alertOnNoData || (alertOnGroupDisappear && groups.length === 0);
if (nextState === AlertStates.NO_DATA && shouldTriggerNoDataAlert) {
```
?

Reply:
Not that I've written this out like this though, I am not sure that's what we want! I've confused myself. 😆 

I think we just want what was there before? How/when does `nextState` equal `AlertStates.NO_DATA`? Will that happen when there are no groups? If so, I don't think we have to check the `alertOnGroupDisappear` or the group length here at all unless I am wildly misunderstanding (likely).

Reply:
Pretty confused by this line as well

Reply:
Incorrect, this handles the case for if there **are** groups, and **`alertOnGroupDisappear` is enabled**, but **`alertOnNoData` is disabled.** Therefore `groups.length > 1`, because there are more groups than `*`.

I'll edit the code to clarify.

Reply:
Code updated, added extensive comments to make sure it's clear what's happening.